src/transactions/transaction_builder/transaction_builder.ts

```diff
@@ -126,6 +131,111 @@ export function generateTransactionPayload(args: GenerateTransactionPayloadData)
   );
 }
 
+export async function generateTransactionPayloadWithRemoteABI(
+  args: EntryFunctionDataWithABI,
+  aptosConfig: AptosConfig,
+): Promise<TransactionPayloadEntryFunction>;
+export async function generateTransactionPayloadWithRemoteABI(
+  args: MultiSigDataWithABI,
+  aptosConfig: AptosConfig,
+): Promise<TransactionPayloadMultisig>;
+export async function generateTransactionPayloadWithRemoteABI(
+  args: GenerateTransactionPayloadDataWithABI,
+  aptosConfig: AptosConfig,
+): Promise<TransactionPayload>;
+
+export async function generateTransactionPayloadWithRemoteABI(
+  args: GenerateTransactionPayloadDataWithABI,
+  aptosConfig: AptosConfig,
+): Promise<TransactionPayload> {
+  const funcNameParts = args.function.split("::");
+  const moduleAddress = funcNameParts[0];
+  const moduleName = funcNameParts[1];
+  const functionName = funcNameParts[2];
+
+  // We fetch the entry function ABI, and then pretend that we already had the ABI
+  const functionAbi = await fetchEntryFunctionAbi(moduleAddress, moduleName, functionName, aptosConfig);
+  return generateTransactionPayloadWithABI(args, functionAbi);
+}
+
+export function generateTransactionPayloadWithABI(
+  args: EntryFunctionDataWithABI,
+  functionAbi: MoveFunction,
+): TransactionPayloadEntryFunction;
+export function generateTransactionPayloadWithABI(
+  args: MultiSigDataWithABI,
+  functionAbi: MoveFunction,
+): TransactionPayloadMultisig;
+export function generateTransactionPayloadWithABI(
+  args: GenerateTransactionPayloadDataWithABI,
+  functionAbi: MoveFunction,
+): TransactionPayload;
+export function generateTransactionPayloadWithABI(
+  args: GenerateTransactionPayloadDataWithABI,
+  functionAbi: MoveFunction,
+): TransactionPayload {
+  const funcNameParts = args.function.split("::");
+  const moduleAddress = funcNameParts[0];
+  const moduleName = funcNameParts[1];
+  const functionName = funcNameParts[2];
+
+  // We have to find the first non-signer arg for arguments
+  const startAbiIndex = findFirstNonSignerArg(functionAbi);
+  let abiIndex = startAbiIndex;
+
+  // Ensure that all type arguments are typed properly
+  const typeArguments = typeTagConversion(args.typeArguments);
+
+  // Check the type argument count against the ABI
+  if (typeArguments.length !== functionAbi?.generic_type_params.length) {
+    throw new Error(
+      `Type argument count mismatch, expected ${functionAbi.generic_type_params.length}, received ${typeArguments.length}`,
+    );
+  }
+
+  // Check all BCS types, and convert any non-BCS types
+  const functionArguments: Array<EntryFunctionArgumentTypes> = [];
+  for (let i = 0; i < args.functionArguments.length; i += 1) {
```

Comment:
args.functionArguments.forEach might be shorter :)

Reply:
Yeah, I'm reworking this so I don't need it, but I do need it indexed (for error messages)